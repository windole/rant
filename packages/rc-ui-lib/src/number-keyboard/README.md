# NumberKeyboard 数字键盘

### 介绍

H5 数字键盘，可以配合密码输入框组件或自定义的输入框组件使用

### 引入

```js
import { NumberKeyboard } from 'rc-ui-lib';
```

## 代码演示

### 默认样式

数字键盘提供了 onInput、onDelete、onBlur、onClose 事件，分别对应输入内容、删除内容失去焦点和关闭的动作。

```jsx
<Cell onClick={setVisible}>弹出默认键盘</Cell>
<NumberKeyboard
  visible={visible}
  onClose={actions.onClose}
  onInput={actions.onInput}
  onDelete={actions.onDelete}
  onBlur={actions.onBlur}
/>
```

```js
const [visible, setVisible] = useState(false);

const actions = {
  onClose: () => {
    Toast.info('closed');
    setVisible('');
  },
  onInput: (key: string) => {
    Toast.info(key);
  },
  onDelete: () => {
    Toast.info('delete');
  },
  onBlur: () => {
    setVisible('');
  },
};
```

### 带右侧栏的键盘

将 theme 属性设置为 `custom` 来展示键盘的右侧栏，常用于输入金额的场景。

```jsx
<NumberKeyboard
  visible={visible}
  theme="custom"
  extraInfo="."
  onClose={actions.onClose}
  onInput={actions.onInput}
  onDelete={actions.onDelete}
  onBlur={actions.onBlur}
/>
```

### 身份证号键盘

通过 `extraKey` 属性可以设置左下角按键内容，比如需要输入身份证号时，可以将 `extraKey` 设置为 `X`。

```jsx
<NumberKeyboard
  visible={visible}
  extraKey="X"
  onClose={actions.onClose}
  onInput={actions.onInput}
  onDelete={actions.onDelete}
  onBlur={actions.onBlur}
/>
```

### 键盘标题

通过 `title` 属性可以设置键盘标题。

```jsx
<NumberKeyboard
  visible={visible}
  title="键盘标题"
  extraKey="."
  closeButtonText="完成"
  onClose={actions.onClose}
  onInput={actions.onInput}
  onDelete={actions.onDelete}
  onBlur={actions.onBlur}
/>
```

### 配置多个按键

当 theme 为 `custom` 时，支持以数组的形式配置两个 `extraKey`。

```jsx
<NumberKeyboard
  visible={visible}
  theme="custom"
  extraKey={['00', '.']}
  closeButtonText="完成"
  onClose={actions.onClose}
  onInput={actions.onInput}
  onDelete={actions.onDelete}
  onBlur={actions.onBlur}
/>
```

### 随机数字键盘

通过 `randomKeyOrder` 属性可以随机排序数字键盘，常用于安全等级较高的场景。

```jsx
<NumberKeyboard
  visible={visible}
  randomKeyOrder
  onClose={actions.onClose}
  onInput={actions.onInput}
  onDelete={actions.onDelete}
  onBlur={actions.onBlur}
/>
```

## API

### Props

| 参数名 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| visible | 是否显示键盘 | _boolean_ | - |
| title | 键盘标题 | _string_ | - |
| theme | 样式风格，可选值为 `custom` | _string_ | `default` |
| transition | 是否开启过场动画 | _boolean_ | `true` |
| zIndex | 键盘 z-index 层级 | \_number | `100` |
| extraKey | 底部额外按键的内容 | _string \| string[]_ | `''` |
| closeButtonText | 关闭按钮文字，空则不展示 | _string_ | - |
| deleteButtonText | 删除按钮文字，空则展示删除图标 | _string_ | - |
| closeButtonLoading | 是否将关闭按钮设置为加载中状态，仅在 `theme="custom"` 时有效 | _boolean_ | `false` |
| showDeleteKey | 是否展示删除图标 | _boolean_ | `true` |
| blurOnClose | 是否在点击关闭按钮时触发 blur 事件 | _boolean_ | `true` |
| hideOnClickOutside | 是否在点击外部时收起键盘 | _boolean_ | `true` |
| safeAreaInsetBottom | 是否开启[底部安全区适配](#/zh-CN/adrcced-usage#di-bu-an-quan-qu-gua-pei) | _boolean_ | `true` |
| randomKeyOrder | 是否将通过随机顺序展示按键 | _boolean_ | `false` |
| titleLeft | 自定义标题栏左侧内容 | React.ReactNode | - |
| teleport | 指定挂载的节点 | _HTMLElement_ | _(() => HTMLElement)_ |

### Events

| 事件名   | 说明                           | 回调参数      |
| -------- | ------------------------------ | ------------- |
| onInput  | 点击按键时触发                 | key: 按键内容 |
| onDelete | 点击删除键时触发               | -             |
| onBlur   | 点击关闭按钮或非键盘区域时触发 | -             |
| onClose  | 点击关闭按钮时触发             | -             |

### 类型定义

组件导出以下类型定义：

```js
import type { NumberKeyboardProps, NumberKeyboardTheme } from 'rc-ui-lib';
```

## 主题定制

### 样式变量

组件提供了下列 CSS 变量，可用于自定义样式，使用方法请参考 [ConfigProvider 组件](#/zh-CN/config-provider)。

| 名称                                         | 默认值                      | 描述 |
| -------------------------------------------- | --------------------------- | ---- |
| --rc-number-keyboard-background-color        | _var(--rc-gray-2)_          | -    |
| --rc-number-keyboard-key-height              | _48px_                      | -    |
| --rc-number-keyboard-key-font-size           | _28px_                      | -    |
| --rc-number-keyboard-key-active-color        | _var(--rc-gray-3)_          | -    |
| --rc-number-keyboard-key-background-color    | _var(--rc-white)_           | -    |
| --rc-number-keyboard-delete-font-size        | _var(--rc-font-size-lg)_    | -    |
| --rc-number-keyboard-title-color             | _var(--rc-gray-7)_          | -    |
| --rc-number-keyboard-title-height            | _34px_                      | -    |
| --rc-number-keyboard-title-font-size         | _var(--rc-font-size-lg)_    | -    |
| --rc-number-keyboard-close-padding           | _0 var(--rc-padding-md)_    | -    |
| --rc-number-keyboard-close-color             | _var(--rc-text-link-color)_ | -    |
| --rc-number-keyboard-close-font-size         | _var(--rc-font-size-md)_    | -    |
| --rc-number-keyboard-button-text-color       | _var(--rc-white)_           | -    |
| --rc-number-keyboard-button-background-color | _var(--rc-primary-color)_   | -    |
| --rc-number-keyboard-z-index                 | _100_                       | -    |
