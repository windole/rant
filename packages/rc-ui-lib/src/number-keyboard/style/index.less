@import './var.less';

:root {
  --rc-number-keyboard-background-color: @number-keyboard-background-color;
  --rc-number-keyboard-key-height: @number-keyboard-key-height;
  --rc-number-keyboard-key-font-size: @number-keyboard-key-font-size;
  --rc-number-keyboard-key-active-color: @number-keyboard-key-active-color;
  --rc-number-keyboard-key-background-color: @number-keyboard-key-background-color;
  --rc-number-keyboard-delete-font-size: @number-keyboard-delete-font-size;
  --rc-number-keyboard-title-color: @number-keyboard-title-color;
  --rc-number-keyboard-title-height: @number-keyboard-title-height;
  --rc-number-keyboard-title-font-size: @number-keyboard-title-font-size;
  --rc-number-keyboard-close-padding: @number-keyboard-close-padding;
  --rc-number-keyboard-close-color: @number-keyboard-close-color;
  --rc-number-keyboard-close-font-size: @number-keyboard-close-font-size;
  --rc-number-keyboard-button-text-color: @number-keyboard-button-text-color;
  --rc-number-keyboard-button-background-color: @number-keyboard-button-background-color;
  --rc-number-keyboard-z-index: @number-keyboard-z-index;
}

.rc-number-keyboard {
  z-index: var(--rc-number-keyboard-z-index);
  width: 100%;
  padding-bottom: 22px;
  background: var(--rc-number-keyboard-background-color);
  user-select: none;

  &--with-title {
    border-radius: 20px 20px 0 0;
  }

  &__header {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: content-box;
    height: var(--rc-number-keyboard-title-height);
    padding-top: 6px;
    color: var(--rc-number-keyboard-title-color);
    font-size: var(--rc-number-keyboard-title-font-size);
  }

  &__title {
    display: inline-block;
    font-weight: normal;

    &-left {
      position: absolute;
      left: 0;
    }
  }

  &__body {
    display: flex;
    padding: 6px 0 0 6px;
  }

  &__keys {
    display: flex;
    flex: 3;
    flex-wrap: wrap;
  }

  &__close {
    position: absolute;
    right: 0;
    height: 100%;
    padding: var(--rc-number-keyboard-close-padding);
    color: var(--rc-number-keyboard-close-color);
    font-size: var(--rc-number-keyboard-close-font-size);
    background-color: transparent;
    border: none;
    cursor: pointer;
  }

  &__sidebar {
    display: flex;
    flex: 1;
    flex-direction: column;
  }

  &--unfit {
    padding-bottom: 0;
  }
}

.rc-key {
  display: flex;
  align-items: center;
  justify-content: center;
  height: var(--rc-number-keyboard-key-height);
  font-size: var(--rc-number-keyboard-key-font-size);
  line-height: 1.5;
  background: var(--rc-number-keyboard-key-background-color);
  border-radius: var(--rc-border-radius-lg);
  cursor: pointer;

  &--large {
    // height: 100% can't fill flex parent on legacy safari
    // see: https://stackoverflow.com/questions/33636796
    position: absolute;
    top: 0;
    right: 6px;
    bottom: 6px;
    left: 0;
    height: auto;
  }

  &--blue,
  &--delete {
    font-size: var(--rc-number-keyboard-delete-font-size);
  }

  &--active {
    background-color: var(--rc-number-keyboard-key-active-color);
  }

  &--blue {
    color: var(--rc-number-keyboard-button-text-color);
    background: var(--rc-number-keyboard-button-background-color);

    &.rc-key--active {
      opacity: var(--rc-active-opacity);
    }
  }

  &__wrapper {
    position: relative;
    flex: 1;
    flex-basis: 33%;
    box-sizing: border-box;
    padding: 0 6px 6px 0;

    &--wider {
      flex-basis: 66%;
    }
  }

  &__delete-icon {
    width: 32px;
    height: 22px;
  }

  &__collapse-icon {
    width: 30px;
    height: 24px;
  }

  &__loading-icon {
    color: var(--rc-number-keyboard-button-text-color);
  }
}
