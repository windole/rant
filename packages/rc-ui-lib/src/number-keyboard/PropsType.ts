import React from 'react';
import { TeleportType } from '../utils';

export type NumberKeyboardTheme = 'default' | 'custom';

export type KeyType = '' | 'delete' | 'extra' | 'close';

export type KeyConfig = {
  text?: number | string;
  type?: KeyType;
  color?: string;
  wider?: boolean;
};

export interface NumberKeyboardProps {
  /** 是否显示键盘 */
  visible?: boolean;
  /** 键盘标题 */
  title?: string;
  /** 样式风格，可选值为 `custom`、`default` */
  theme?: NumberKeyboardTheme;
  /** 键盘 z-index 层级 */
  zIndex?: number;
  /**
   * 是否在点击关闭按钮时触发 blur 事件
   * 默认值为`true`
   */
  blurOnClose?: boolean;
  /**
   * 是否展示删除图标
   * 默认值为`true`
   */
  showDeleteKey?: boolean;
  /** 是否将通过随机顺序展示按键 */
  randomKeyOrder?: boolean;
  /** 关闭按钮文字，空则不展示 */
  closeButtonText?: string;
  /** 删除按钮文字，空则展示删除图标 */
  deleteButtonText?: string;
  /**
   * 是否将关闭按钮设置为加载中状态，
   * 仅在 `theme="custom"` 时有效
   */
  closeButtonLoading?: boolean;
  /**
   * 是否在点击外部时收起键盘
   * 默认值为`true`
   */
  hideOnClickOutside?: boolean;
  /**
   * 是否开启底部安全区适配
   * 默认值为`true`
   */
  safeAreaInsetBottom?: boolean;
  /** 底部额外按键的内容 */
  extraKey?: string | string[];
  /**
   * 自定义标题左侧
   */
  titleLeft?: React.ReactNode;
  /**
   * 点击关闭按钮时触发
   */
  onClose?: () => void;
  /** 点击按键时触发 */
  onInput?: (v: string) => void;
  /** 点击删除键时触发 */
  onDelete?: () => void;
  /** 点击关闭按钮或非键盘区域时触发 */
  onBlur?: () => void;
  /** 指定挂载的节点 */
  teleport?: TeleportType;
}
