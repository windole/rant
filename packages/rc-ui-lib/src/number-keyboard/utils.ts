import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
