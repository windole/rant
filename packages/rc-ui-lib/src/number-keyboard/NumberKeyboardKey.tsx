import React, { useContext, useState } from 'react';
import classNames from 'classnames';
import Loading from '../loading';

import { useTouch } from '../hooks/use-touch';

import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { KeyType } from './PropsType';
import { preventDefault } from '../utils';

interface NumberKeyboardKeyProps {
  type: KeyType;
  text: string | number;
  color?: string;
  wider?: boolean;
  large?: boolean;
  loading?: boolean;
  onPress: (text: string, type: KeyType) => void;
}

const NumberKeyboardKey: React.FC<NumberKeyboardKeyProps> = (props) => {
  const { type, text, color, wider, large, loading, onPress, children } = props;
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('key', prefixCls);

  const touch = useTouch();
  const [active, setActive] = useState(false);

  const CollapseIcon = (
    <svg className={classNames(bem('collapse-icon'))} viewBox="0 0 30 24">
      <path
        d="M26 13h-2v2h2v-2zm-8-3h2V8h-2v2zm2-4h2V4h-2v2zm2 4h4V4h-2v4h-2v2zm-7 14 3-3h-6l3 3zM6 13H4v2h2v-2zm16 0H8v2h14v-2zm-12-3h2V8h-2v2zM28 0l1 1 1 1v15l-1 2H1l-1-2V2l1-1 1-1zm0 2H2v15h26V2zM6 4v2H4V4zm10 2h2V4h-2v2zM8 9v1H4V8zm8 0v1h-2V8zm-6-5v2H8V4zm4 0v2h-2V4z"
        fill="currentColor"
      />
    </svg>
  );

  const DeleteIcon = (
    <svg className={classNames(bem('delete-icon'))} viewBox="0 0 32 22">
      <path
        d="M28 0a4 4 0 0 1 4 4v14a4 4 0 0 1-4 4H10.4a2 2 0 0 1-1.4-.6L1 13.1c-.6-.5-.9-1.3-.9-2 0-1 .3-1.7.9-2.2L9 .6a2 2 0 0 1 1.4-.6zm0 2H10.4l-8.2 8.3a1 1 0 0 0-.3.7c0 .3.1.5.3.7l8.2 8.4H28a2 2 0 0 0 2-2V4c0-1.1-.9-2-2-2zm-5 4a1 1 0 0 1 .7.3 1 1 0 0 1 0 1.4L20.4 11l3.3 3.3c.2.2.3.5.3.7 0 .3-.1.5-.3.7a1 1 0 0 1-.7.3 1 1 0 0 1-.7-.3L19 12.4l-3.4 3.3a1 1 0 0 1-.6.3 1 1 0 0 1-.7-.3 1 1 0 0 1-.3-.7c0-.2.1-.5.3-.7l3.3-3.3-3.3-3.3A1 1 0 0 1 14 7c0-.3.1-.5.3-.7A1 1 0 0 1 15 6a1 1 0 0 1 .6.3L19 9.6l3.3-3.3A1 1 0 0 1 23 6z"
        fill="currentColor"
      />
    </svg>
  );

  const onTouchStart = (event) => {
    touch.start(event);
    setActive(true);
  };

  const onTouchMove = (event) => {
    touch.move(event);

    if (touch.direction.current) {
      setActive(false);
    }
  };

  const onTouchEnd = (event) => {
    if (active) {
      // eliminate tap delay on safari
      if (!children) {
        preventDefault(event);
      }
      setActive(false);
      onPress(String(text), type);
    }
  };

  const renderContent = () => {
    if (loading) {
      return <Loading className={classNames(bem('loading-icon'))} />;
    }
    const contentText = children || text;
    switch (type) {
      case 'delete':
        return contentText || DeleteIcon;
      case 'extra':
        return contentText || CollapseIcon;
      default:
        return contentText;
    }
  };

  return (
    <div
      className={classNames(bem('wrapper', { wider }))}
      onTouchStart={onTouchStart}
      onTouchMove={onTouchMove}
      onTouchEnd={onTouchEnd}
      onTouchCancel={onTouchEnd}
    >
      <div
        role="button"
        tabIndex={0}
        className={classNames(
          bem([
            color,
            {
              large,
              active,
              delete: type === 'delete',
            },
          ]),
        )}
      >
        {renderContent()}
      </div>
    </div>
  );
};

NumberKeyboardKey.defaultProps = {
  color: '',
  wider: false,
  large: false,
  loading: false,
};

export default NumberKeyboardKey;
