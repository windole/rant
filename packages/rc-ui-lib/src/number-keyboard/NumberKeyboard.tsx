import React, { useContext, useMemo, useRef } from 'react';
import classNames from 'classnames';
import { Popup } from '../popup';
import { NumberKeyboardProps, KeyType } from './PropsType';

import { genCustomKeys, genDefaultKeys } from './utils';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import NumberKeyboardKey from './NumberKeyboardKey';
import useClickAway from '../hooks/use-click-away';

const NumberKeyboard: React.FC<NumberKeyboardProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('number-keyboard', prefixCls);
  const {
    visible,
    safeAreaInsetBottom,
    theme,
    blurOnClose,
    showDeleteKey,
    hideOnClickOutside,
    zIndex,
    teleport,
    title,
    closeButtonText,
    titleLeft,
  } = props;

  const root = useRef<HTMLDivElement>(null);

  const showClose = closeButtonText && theme === 'default';
  const showTitle = !!(title || showClose || titleLeft);

  const keys = useMemo(
    () => (theme === 'custom' ? genCustomKeys(props) : genDefaultKeys(props)),
    [theme],
  );

  const onBlur = () => {
    if (visible) {
      props.onBlur?.();
    }
  };

  const onClose = () => {
    props.onClose?.();

    if (blurOnClose) {
      onBlur();
    }
  };

  const onPress = (text: string, type: KeyType) => {
    if (text === '') {
      if (type === 'extra') {
        onBlur();
      }
      return;
    }

    if (type === 'delete') {
      props.onDelete?.();
    } else if (type === 'close') {
      onClose();
    } else {
      props.onInput?.(text);
    }
  };

  const renderTitle = () => {
    if (!showTitle) {
      return null;
    }
    return (
      <div className={classNames(bem('header'))}>
        {titleLeft && <span className={classNames(bem('title-left'))}>{titleLeft}</span>}
        {title && <h2 className={classNames(bem('title'))}>{title}</h2>}
        {showClose && (
          <button type="button" className={classNames(bem('close'))} onClick={onClose}>
            {closeButtonText}
          </button>
        )}
      </div>
    );
  };

  const renderKeys = () => {
    return keys.map((key) => {
      return (
        <NumberKeyboardKey
          key={`${key.text}${key.type}`}
          text={key.text}
          type={key.type}
          wider={key.wider}
          color={key.color}
          onPress={onPress}
        />
      );
    });
  };

  const renderSidebar = () => {
    if (theme === 'custom') {
      return (
        <div className={classNames(bem('sidebar'))}>
          {showDeleteKey && (
            <NumberKeyboardKey
              large
              text={props.deleteButtonText}
              type="delete"
              onPress={onPress}
            />
          )}
          <NumberKeyboardKey
            large
            text={props.closeButtonText}
            type="close"
            color="blue"
            loading={props.closeButtonLoading}
            onPress={onPress}
          />
        </div>
      );
    }
    return null;
  };

  useClickAway(root, onBlur, 'touchstart', hideOnClickOutside);

  return (
    <Popup
      visible={visible}
      preventDefaultMouseDown
      overlay={false}
      round={showTitle}
      zIndex={zIndex}
      position="bottom"
      duration={300}
      teleport={teleport}
      safeAreaInsetBottom={safeAreaInsetBottom}
      onClose={onClose}
    >
      <div ref={root} className={classNames(bem())}>
        {renderTitle()}
        <div className={classNames(bem('body'))}>
          <div className={classNames(bem('keys'))}>{renderKeys()}</div>
          {renderSidebar()}
        </div>
      </div>
    </Popup>
  );
};

NumberKeyboard.defaultProps = {
  theme: 'default',
  visible: false,
  extraKey: '',
  showDeleteKey: true,
  zIndex: 100,
  closeButtonLoading: false,
  hideOnClickOutside: true,
  randomKeyOrder: false,
  safeAreaInsetBottom: true,
  blurOnClose: true,
};

export default NumberKeyboard;
