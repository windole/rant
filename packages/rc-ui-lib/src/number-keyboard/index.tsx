import NumberKeyboard from './NumberKeyboard';

export default NumberKeyboard;
export { NumberKeyboard };
export type { NumberKeyboardProps, NumberKeyboardTheme } from './PropsType';
