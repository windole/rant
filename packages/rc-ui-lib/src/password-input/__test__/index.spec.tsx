import React, { useRef } from 'react';
import { fireEvent, render, cleanup, waitFor } from '@testing-library/react';
import { PasswordInput, PasswordInputInstance, PasswordInputProps } from '..';
import { NumberKeyboard } from '../../number-keyboard';
import TestsEvent from '../../../tests/events';

const clickKey = async (target) => {
  await TestsEvent.triggerTouch(target, 'touchstart', [[0, 0]]);
  await TestsEvent.triggerTouch(target, 'touchend', [[0, 0]]);
};

const $props = {
  focused: false,
};

describe('PasswordInput', () => {
  function createPasswordInput(props: PasswordInputProps) {
    const { baseElement, getAllByRole, getByText, container, rerender, debug } = render(
      <PasswordInput {...props} />,
    );

    return {
      baseElement,
      container,
      rerender,
      getByText,
      getAllByRole,
      debug,
    };
  }

  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  // it('should emit onFocus event when security is touched', async () => {
  //   const onFocus = jest.fn();
  //   const props = {
  //     ...$props,
  //     onFocus,
  //   };

  //   const { container } = createPasswordInput(props);
  //   const wrapper = container.querySelector('.rc-password-input__security');
  //   await TestsEvent.triggerTouch(wrapper, 'touchstart', [[0, 0]]);
  //   expect(onFocus).toHaveBeenCalled();
  // });

  it('should emit onFocus event when wrapper is focused', async () => {
    const onFocus = jest.fn();
    const props = {
      ...$props,
      onFocus,
    };

    const { container } = createPasswordInput(props);
    const wrapper = container.querySelector('.rc-password-input');
    await fireEvent.focus(wrapper);
    expect(onFocus).toHaveBeenCalled();
  });

  it('should emit onBlur event after clicking outside', async () => {
    const onBlur = jest.fn();
    const props = {
      ...$props,
      focused: true,
      keyboard: <NumberKeyboard theme="custom" />,
      onBlur,
    };

    const { getAllByRole, container } = createPasswordInput(props);
    const keyBlur = getAllByRole('button')[9];
    const keyClose = getAllByRole('button')[12];
    const wrapper = container.querySelector('.rc-password-input');

    await clickKey(keyBlur);
    await clickKey(keyClose);
    await fireEvent.blur(wrapper);

    expect(onBlur).toHaveBeenCalled();
  });

  it('should emit onChange event after input number', async () => {
    const onChange = jest.fn();
    const props = {
      ...$props,
      focused: true,
      onChange,
      keyboard: <NumberKeyboard />,
    };

    const { getAllByRole } = createPasswordInput(props);
    const key = getAllByRole('button')[0];
    await clickKey(key);
    expect(onChange).toHaveBeenCalledWith('1');
  });

  it('should emit onChange event after press delete button', async () => {
    const onChange = jest.fn();
    const props = {
      ...$props,
      value: '12',
      focused: true,
      onChange,
      keyboard: <NumberKeyboard />,
    };

    const { getAllByRole } = createPasswordInput(props);
    const key = getAllByRole('button')[11];
    await clickKey(key);
    expect(onChange).toHaveBeenCalledWith('1');
  });

  it('should emit onFill event after input number', async () => {
    const onFill = jest.fn();
    const props = {
      ...$props,
      focused: true,
      onFill,
      keyboard: <NumberKeyboard />,
    };

    const { getAllByRole } = createPasswordInput(props);
    const key = getAllByRole('button');
    await clickKey(key[0]);
    await clickKey(key[1]);
    await clickKey(key[2]);
    await clickKey(key[3]);
    await clickKey(key[4]);
    await clickKey(key[5]);
    expect(onFill).toHaveBeenCalledWith('123456');
  });

  it('should render error info correctly', async () => {
    const props = {
      ...$props,
      errorInfo: 'error!',
    };

    const { container } = createPasswordInput(props);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when using mask prop', async () => {
    const props = {
      ...$props,
      value: '1',
      mask: false,
    };

    const { getByText } = createPasswordInput(props);
    const item = getByText('1');
    expect(item).toMatchSnapshot();
  });

  it('should render correctly when using gutter prop', async () => {
    const props = {
      ...$props,
      gutter: 10,
    };

    const { container } = createPasswordInput(props);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when using length prop', async () => {
    const props = {
      ...$props,
      length: 4,
    };

    const { container, rerender } = createPasswordInput(props);
    expect(container).toMatchSnapshot();
    const props2 = {
      ...$props,
      length: 6,
      value: undefined,
    };
    await waitFor(() => rerender(<PasswordInput {...props2} />));
    expect(container).toMatchSnapshot();
  });

  it("test instance's function", async () => {
    const onBlur = jest.fn();
    const onFocus = jest.fn();
    const onChange = jest.fn();
    const Demo = () => {
      const passwordInputRef = useRef<PasswordInputInstance>(null);

      const handleFocus = () => {
        passwordInputRef.current?.focus();
      };

      const handleBlur = () => {
        passwordInputRef.current?.blur();
      };
      const props = {
        ...$props,
        onFocus,
        onBlur,
        onChange,
      };

      return (
        <div>
          <PasswordInput ref={passwordInputRef} {...props} />,
          <button type="button" data-testid="button1" onClick={handleFocus}>
            handleFocus
          </button>
          <button type="button" data-testid="button2" onClick={handleBlur}>
            handleBlur
          </button>
        </div>
      );
    };

    const { container, getByTestId } = render(<Demo />);

    const buttonBox1 = getByTestId('button1');
    const buttonBox2 = getByTestId('button2');

    await fireEvent.click(buttonBox1);
    expect(onFocus).toHaveBeenCalled();

    const input = container.querySelector('.rc-password-input__native-input');
    fireEvent.change(input, { target: { value: '123' } });
    expect(onChange).toHaveBeenCalledWith('123');

    fireEvent.click(buttonBox2);
    expect(onBlur).toHaveBeenCalled();
  });
});
