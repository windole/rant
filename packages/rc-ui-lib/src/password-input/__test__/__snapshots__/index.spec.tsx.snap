// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput should render correctly when using gutter prop 1`] = `
<div>
  <div
    class="rc-password-input"
    tabindex="0"
  >
    <ul
      class="rc-password-input__security"
    >
      <li
        class="rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-password-input__item"
        style="margin-left: 10px;"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-password-input__item"
        style="margin-left: 10px;"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-password-input__item"
        style="margin-left: 10px;"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-password-input__item"
        style="margin-left: 10px;"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-password-input__item"
        style="margin-left: 10px;"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
    </ul>
    <input
      class="rc-password-input__native-input"
      inputmode="numeric"
      pattern="[0-9]*"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`PasswordInput should render correctly when using length prop 1`] = `
<div>
  <div
    class="rc-password-input"
    tabindex="0"
  >
    <ul
      class="rc-password-input__security rc-hairline--surround"
    >
      <li
        class="rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
    </ul>
    <input
      class="rc-password-input__native-input"
      inputmode="numeric"
      pattern="[0-9]*"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`PasswordInput should render correctly when using length prop 2`] = `
<div>
  <div
    class="rc-password-input"
    tabindex="0"
  >
    <ul
      class="rc-password-input__security rc-hairline--surround"
    >
      <li
        class="rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
    </ul>
    <input
      class="rc-password-input__native-input"
      inputmode="numeric"
      pattern="[0-9]*"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`PasswordInput should render correctly when using mask prop 1`] = `
<li
  class="rc-password-input__item"
>
  1
</li>
`;

exports[`PasswordInput should render error info correctly 1`] = `
<div>
  <div
    class="rc-password-input"
    tabindex="0"
  >
    <ul
      class="rc-password-input__security rc-hairline--surround"
    >
      <li
        class="rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
      <li
        class="rc-hairline--left rc-password-input__item"
      >
        <i
          style="visibility: hidden;"
        />
      </li>
    </ul>
    <div
      class="rc-password-input__error-info"
    >
      error!
    </div>
    <input
      class="rc-password-input__native-input"
      inputmode="numeric"
      pattern="[0-9]*"
      type="text"
      value=""
    />
  </div>
</div>
`;
